cibuildwheel/macos.py

```diff
@@ -50,19 +52,42 @@ def call(args, env=None, cwd=None, shell=False):
-        if python_package_identifier not in installed_system_packages:
-            # download the pkg
-            call(['curl', '-L', '-o', '/tmp/Python.pkg', config.url])
-            # install
-            call(['sudo', 'installer', '-pkg', '/tmp/Python.pkg', '-target', '/'])
-            # patch open ssl
-            if config.version == '3.5':
-                call(['curl', '-fsSLo', '/tmp/python-patch.tar.gz', 'https://github.com/mayeut/patch-macos-python-openssl/releases/download/v1.0.2t/patch-macos-python-%s-openssl-v1.0.2t.tar.gz' % config.version])
-                call(['sudo', 'tar', '-C', '/Library/Frameworks/Python.framework/Versions/%s/' % config.version, '-xmf', '/tmp/python-patch.tar.gz'])
-
-        installation_bin_path = '/Library/Frameworks/Python.framework/Versions/{}/bin'.format(config.version)
+        if config.identifier.startswith('cp'):
+            # if this version of python isn't installed, get it from python.org and install
+            python_package_identifier = 'org.python.Python.PythonFramework-%s' % config.version
+            if python_package_identifier not in installed_system_packages:
+                # download the pkg
+                call(['curl', '-L', '-o', '/tmp/Python.pkg', config.url])
+                # install
+                call(['sudo', 'installer', '-pkg', '/tmp/Python.pkg', '-target', '/'])
+                # patch open ssl
+                if config.version == '3.5':
+                    call(['curl', '-fsSLo', '/tmp/python-patch.tar.gz', 'https://github.com/mayeut/patch-macos-python-openssl/releases/download/v1.0.2t/patch-macos-python-%s-openssl-v1.0.2t.tar.gz' % config.version])
+                    call(['sudo', 'tar', '-C', '/Library/Frameworks/Python.framework/Versions/%s/' % config.version, '-xmf', '/tmp/python-patch.tar.gz'])
+
+            installation_bin_path = '/Library/Frameworks/Python.framework/Versions/{}/bin'.format(config.version)
+        elif config.identifier.startswith('pp'):
+            pypy_tar_bz2 = config.url.rsplit('/', 1)[-1]
+            assert pypy_tar_bz2.endswith(".tar.bz2")
+            pypy_base_filename = os.path.splitext(os.path.splitext(pypy_tar_bz2)[0])[0]
+            installation_bin_path = os.path.join('/tmp', pypy_base_filename, 'bin')
+            if not os.path.exists(installation_bin_path):
+                call(['curl', '-L', '-o', os.path.join("/tmp", pypy_tar_bz2), config.url])
+                call(['tar', '-C', '/tmp', '-xf', os.path.join("/tmp", pypy_tar_bz2)])
+                pypy_executable = 'pypy' if config.version[0] == '2' else 'pypy3'
+                python_symlink = 'python' if config.version[0] == '2' else 'python3'
+                os.symlink(os.path.join(installation_bin_path, pypy_executable), os.path.join(installation_bin_path, python_symlink))
+                call([os.path.join(installation_bin_path, pypy_executable), '-m', 'ensurepip'])
```

Comment:
@joerick, @mayeut This is quite a mess. Especially because afterwards, new symlinks get created in `/tmp/cibw_bin` to get rid of `python`/`python3` differences in executable names.

Any ideas on how to improve this?

Reply:
I really don't know how to improve this... All other solutions I'm seeing are as messy as the current situation.